### DISCUSSION TITLE
chapter_hashing/summary/

### DISCUSSION THREAD

**Discussion by giscus** [krahets/hello-algo#100] [Announcements]
# chapter_hashing/summary/

Your first book to learn Data Structure And Algorithm.

https://www.hello-algo.com/chapter_hashing/summary/

**Comment by 1ghosts** [comment]
最后，哈希表的时间复杂度可能发生劣化。例如在链式地址中，我们采取在链表或红黑树中执行查找操作，仍然有退化至 O(n)
 时间的风险。
如果是红黑树（恒定logN），这是因为hashcode都到同一个桶里才导致的吧和红黑树无关吧

**Comment by swssws333** [comment]
最后，哈希表的时间复杂度可能发生劣化。例如在链式地址中，我们采取在链表或红黑树中执行查找操作，仍然有退化至 O(n)

我不明白为什么是O(n)，最坏情况下就应该是O(log n)才对。当哈希冲突都映射到同一个位置的时候，底层是链表，但是链表的长度大于8的时候，就是红黑树恒为O(log n)，除非它不切换为红黑树。
还没有切换为红黑树之前用链表的时候，其实也就是算O(1)，因为它最多也就遍历8次

**Comment by otohey** [comment]
有没有在哈希冲突之后，继续挂一个哈希表而不是挂链表或者红黑树呢？

**Comment by kizux1** [comment]
没有双重哈希(double hashing)吗

**Comment by qzl1-stack** [comment]
？？？？
